internal/encryption/keyprovider.go

```diff
@@ -10,61 +10,132 @@ import (
-// are invalid.
-func (e *targetBuilder) setupKeyProviders() hcl.Diagnostics {
+// valueMap is a helper type for building hcl.EvalContexts for key_providers.
+type valueMap map[string]map[string]cty.Value
+
+func (v valueMap) set(first string, second string, value cty.Value) {
+	if _, ok := v[first]; !ok {
+		v[first] = make(map[string]cty.Value)
+	}
+	v[first][second] = value
+}
+
+func (v valueMap) has(first string, second string) bool {
+	s, ok := v[first]
+	if !ok {
+		return false
+	}
+	_, ok = s[second]
+	return ok
+}
+
+func (v valueMap) hclEvalContext(root string) *hcl.EvalContext {
+	mMap := make(map[string]cty.Value)
+	for name, ms := range v {
+		mMap[name] = cty.ObjectVal(ms)
+	}
+
+	return &hcl.EvalContext{
+		Variables: map[string]cty.Value{root: cty.ObjectVal(mMap)},
+	}
+}
+
+// Given a set of hcl.Traversals, determine the required key provider configs and non-key_provider references
+func filterKeyProviderReferences(cfg *config.EncryptionConfig, deps []hcl.Traversal) ([]config.KeyProviderConfig, []*addrs.Reference, hcl.Diagnostics) {
 	var diags hcl.Diagnostics
 
-	e.keyValues = make(map[string]map[string]cty.Value)
+	var keyProviderDeps []config.KeyProviderConfig
+	// lang.References is going to fail parsing key_provider deps
+	// so we filter them out in nonKeyProviderDeps.
+	var nonKeyProviderDeps []hcl.Traversal
 
-	kpMap := make(map[string]cty.Value)
-	for _, keyProviderConfig := range e.cfg.KeyProviderConfigs {
-		diags = append(diags, e.setupKeyProvider(keyProviderConfig, nil)...)
-		if diags.HasErrors() {
-			return diags
+	// Setting up key providers from deps.
+	for _, dep := range deps {
+		// Key Provider references should be in the form key_provider.type.name
+		if len(dep) != 3 { //nolint:mnd // linting
+			nonKeyProviderDeps = append(nonKeyProviderDeps, dep)
+			continue
 		}
-		for name, kps := range e.keyValues {
-			kpMap[name] = cty.ObjectVal(kps)
+
+		//nolint:errcheck // This will always be a TraverseRoot, panic is OK if that's not the case
+		depRoot := (dep[0].(hcl.TraverseRoot)).Name
+		if depRoot != "key_provider" {
+			nonKeyProviderDeps = append(nonKeyProviderDeps, dep)
+			continue
+		}
+		depTypeAttr, typeOk := dep[1].(hcl.TraverseAttr)
+		depNameAttr, nameOk := dep[2].(hcl.TraverseAttr)
+
+		if !typeOk || !nameOk {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid Key Provider expression format",
+				Detail:   "Expected key_provider.<type>.<name>",
+				Subject:  dep.SourceRange().Ptr(),
+			})
+			continue
 		}
-		e.ctx.Variables["key_provider"] = cty.ObjectVal(kpMap)
+
+		depType := depTypeAttr.Name
+		depName := depNameAttr.Name
+
+		kpc, ok := cfg.GetKeyProvider(depType, depName)
+		if !ok {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Undefined Key Provider",
+				Detail:   fmt.Sprintf("Key provider %s.%s is missing from the encryption configuration.", depType, depName),
```

Comment:
```suggestion
				Summary:  "Reference to undeclared key provider",
				Detail:   fmt.Sprintf("There is no key_provider %q %q block declared in the encryption block.", depType, depName),
```

To make this more consistent with [the similar message for resource references](https://github.com/opentofu/opentofu/blob/35368d990955f225d0fd72ba6409563c992733e1/internal/tofu/evaluate.go#L658-L664).